### PR TITLE
update lecture&course of course8

### PR DESCRIPTION
**Pull request by DawnMagnet**

### CODE REVIEW

File: course8/lec8_en.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
+---
+marp: true
+math: mathjax
+paginate: true
+backgroundImage: url('../pics/background_moonbit.png')
+headingDivider: 1
+---
+
+# Modern Programming Concepts
```
Comment (skylee03) [inline course8/lec8_en.md]: The title should be "Modern Programming Ideology".

File: course8/course_en.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
+# Modern Programming Ideology: Implementing Queues with Mutable Data Structures
+
+## Overview
+
+In modern programming concepts, a queue is an important data structure that follows the First-In-First-Out (FIFO) principle. This course will show you how to use mutable data structures to implement queues, focusing on two implementation methods: circular queues and singly linked lists, and exploring the concepts of tail call and tail recursion.
```
Comment (peter-jerry-ye) [inline course8/course_en.md]:
```suggestion
In modern programming, a queue is an important data structure that follows the First-In-First-Out (FIFO) principle. This course will show you how to use mutable data structures to implement queues, focusing on two implementation methods: circular queues and singly linked lists, and exploring the concepts of tail call and tail recursion.
```

File: course8/course_en.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+### Expanding the Queue
+
+When the number of elements exceeds the length of the array, an expansion operation is required:
+
+```moonbit no-check
+fn push(self: Queue, t: Int) -> Queue {
+  if self.length == self.array.length() {
+    let new_array: Array[Int] = Array::make(self.array.length() * 2, 0)
+    let mut i = 0
+    while i < self.array.length(), i = i + 1 {
```
Comment (peter-jerry-ye) [inline course8/course_en.md]:
```suggestion
    while i < self.array.length() {
```

File: course8/course_en.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      new_array[i] = self.array[(self.start + i) % self.array.length()]
+    }
```
Comment (peter-jerry-ye) [inline course8/course_en.md:66]:
```suggestion
      i = i + 1
    }
```

File: course8/course_en.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  self.start = (self.start + 1) % self.array.length()
+  self.length = self.length - 1
+  self
+}
+```
+
+### Maintaining the Length of the Queue
+
+```moonbit no-check
+fn length(self: Queue) -> Int {
```
Comment (peter-jerry-ye) [inline course8/course_en.md:91]:
```suggestion
```moonbit
```

File: course8/course_en.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+fn make[T]() -> Queue[T] {
+  {
+    array: Array::make(5, T::default()), // Initialize the array with the default value of the type
+    start: 0,
+    end: 0,
+    length: 0
+  }
+}
+```
+
```
Comment (peter-jerry-ye) [inline course8/course_en.md:116]: Content missing

File: course8/course_en.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+## Implementation of Singly Linked Lists
+
+A singly linked list consists of a series of nodes, each containing data and a reference to the next node.
+
+### Definition of Nodes and Linked Lists
+
+```moonbit
+struct Node[T] {
+  val : T
+  mut next : Option[Node[T]]
```
Comment (peter-jerry-ye) [inline course8/course_en.md:128]: Explanation

File: course8/course_en.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+struct LinkedList[T] {
+  mut head : Option[Node[T]]
```
Comment (peter-jerry-ye) [inline course8/course_en.md:132]: Explanation

File: course8/course_en.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+  mut tail : Option[Node[T]]
+}
+```
+
+### Adding Elements
+
```
Comment (peter-jerry-ye) [inline course8/course_en.md:142]: Explanation. Don't just show code.

File: course8/course_en.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  aux(self.head)
+}
+```
+
+### Stack Overflow Problem
+
+When the linked list is too long, the recursive calculation of the length can cause a stack overflow. To solve this problem, we can use tail call optimization.
+
```
Comment (peter-jerry-ye) [inline course8/course_en.md:193]: There's demo in the course video. You should at least give some screenshots here.

File: course8/course_en.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+### Tail Calls and Tail Recursion
+
+A tail call is when the last operation of a function is a function call, and tail recursion is when the last operation is a recursive call to the function itself. Using tail calls can prevent stack overflow problems.
```
Comment (peter-jerry-ye) [inline course8/course_en.md]: > Using tail calls can prevent stack overflow problems.

Why? Explain.

File: course8/course_en.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+  fn aux2(node: Option[Node[T]], cumul: Int) -> Int {
+    match node {
+      None => cumul
+      Some(node) => aux2(node.next, 1 + cumul)
+    }
+  }
+  aux2(self.head, 0)
+}
+```
+
```
Comment (peter-jerry-ye) [inline course8/course_en.md:214]: Also demo here.

File: course8/course_en.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+```moonbit expr
+let a: Array[Int] = Array::make(5, 0)
+a[0] = 1
+a[1] = 2
+println(a) // Output: [1, 2, 0, 0, 0]
+```
+
+### Simple Implementation of Circular Queues
+
+```moonbit no-check
```
Comment (peter-jerry-ye) [inline course8/course_en.md]:
```suggestion
```moonbit
```

File: course8/course_en.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+    self.end = self.array.length()
+    self.array = new_array
+  }
+  self.push(t) // Recursive call to complete the element addition
+}
+```
+
+### Removing Elements
+
+```moonbit no-check
```
Comment (peter-jerry-ye) [inline course8/course_en.md]:
```suggestion
```moonbit
```

File: course8/course_en.md [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+  self.length = self.length + 1
+  self
+}
+```
+
+### Expanding the Queue
+
+When the number of elements exceeds the length of the array, an expansion operation is required:
+
+```moonbit no-check
```
Comment (peter-jerry-ye) [inline course8/course_en.md]:
```suggestion
```moonbit expr
```

**Review by peter-jerry-ye — CHANGES_REQUESTED** [review]
I reinstate that lecture should let people understand when reading it alone. There are few explanations. Not to mention that some content and all the illustrations are missing.

File: course8/course_en.md (hunk 13 above)
Comment (DawnMagnet) [reply course8/course_en.md]: there is queue inside std so can't pass check

File: course8/course_en.md (hunk 15 above)
Comment (DawnMagnet) [reply course8/course_en.md]: the same reason before

**Comment by DawnMagnet** [comment]
Can you review again?

File: course8/course_en.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+struct LinkedList[T] {
+  mut head : Option[Node[T]]
+  mut tail : Option[Node[T]]
+}
+```
+
+`Node[T]` is a generic struct that represents a node in a linked list. It has two fields: `val` and `next`. The `val` field is used to store the value of the node, and its type is `T`, which can be any valid data type. The `next` field represents the reference to the next node in the linked list. It is an optional field that can either hold a reference to the next node or be empty (`None`), indicating the end of the linked list.
+
+`LinkedList[T]` is a generic struct that represents a linked list. It has two mutable fields: `head` and `tail`. The `head` field represents the reference to the first node (head) of the linked list and is initially set to `None` when the linked list is empty. The `tail` field represents the reference to the last node (tail) of the linked list and is also initially set to `None`. The presence of the `tail` field allows for efficient appending of new nodes to the end of the linked list.
+
```
Comment (peter-jerry-ye) [inline course8/course_en.md:140]: And the illustration?

File: course8/course_en.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+  aux(self.head)
+}
+```
+
+### Stack Overflow Problem
+
+To address the stack overflow issue, the concept of tail calls and tail recursion is introduced. A tail call is a function call that is the last operation in a function, and tail recursion is a specific case where the function calls itself as the last operation.
+
+![overflow](overflow.png)
```
Comment (peter-jerry-ye) [inline course8/course_en.md:192]: What are you trying to explain from just a code snippet?

**Review by peter-jerry-ye — CHANGES_REQUESTED** [review]

**Pull request merged**